include/dlaf/matrix/hdf5.h

```diff
@@ -0,0 +1,274 @@
+  };
+  for_each_local_map_with_dataset<T>(matrix.distribution(), dataset, std::move(tile_rw),
+                                     std::move(dataset_read));
+}
+
+template <class T>
+void to_dataset(dlaf::Matrix<const T, Device::CPU>& matrix, const H5::DataSet& dataset) {
+  auto tile_ro = [&matrix](LocalTileIndex ij) { return matrix.read(ij); };
+  auto dataset_write = [&dataset](auto&&... args) {
+    dataset.write(std::forward<decltype(args)>(args)...);
+  };
+  for_each_local_map_with_dataset<T>(matrix.distribution(), dataset, std::move(tile_ro),
+                                     std::move(dataset_write));
+}
+
+}
+
+class FileHDF5 final {
+public:
+  /// File access modes:
+  /// - READONLY, the file will be opened (if should already exist)
+  /// - READWRITE, the file will be created (if should not exist)
+  enum class FileMode {
+    READONLY,
+    READWRITE,
+  };
+
+  /// Create/open a local file.
+  ///
+  /// @p filepath filepath where the file to be opened is or where it will be created
+  /// @p mode file access mode (see FileMode for more details)
+  ///
+  /// @post if mode == READWRITE, being a local file it will not support parallel-write
+  FileHDF5(const std::string& filepath, const FileMode& mode) {
+    file_ = H5::H5File(filepath, mode2flags(mode));
+  }
+
+  /// Create a file that, besides read, supports writing in parallel from different ranks.
+  ///
+  /// @p comm Communicator grouping all ranks that will be able to write in parallel to the file
+  /// @p filepath filepath where the file will be created
+  ///
+  /// @pre @p filepath should not exist
+  /// @post file created will support parallel-write
+  FileHDF5(comm::Communicator comm, const std::string& filepath) {
+    H5::FileAccPropList fapl;
+    DLAF_ASSERT(H5Pset_fapl_mpio(fapl.getId(), comm, MPI_INFO_NULL) >= 0, "Problem setting up MPI-IO.");
+    file_ = H5::H5File(filepath, mode2flags(FileMode::READWRITE), {}, fapl);
+    has_mpio_ = true;
+    rank_ = comm.rank();
+  }
+
+  /// Write @p matrix into dataset @p dataset_name
+  ///
+  /// @pre if @p matrix is distributed, the file should support parallel-write
+  template <class T, Device D>
+  void write(matrix::Matrix<const T, D>& matrix, const std::string& dataset_name) const {
+    matrix::MatrixMirror<const T, Device::CPU, D> matrix_mirror(matrix);
+
+    matrix::Matrix<const T, Device::CPU>& matrix_host = matrix_mirror.get();
+
+    const bool is_local_matrix = matrix::local_matrix(matrix_host);
+
+    DLAF_ASSERT(is_local_matrix || has_mpio_,
+                "You are trying to store a distributed matrix using a local only file", is_local_matrix,
+                has_mpio_);
+
+    const hsize_t dims_file[3] = {
+        dlaf::to_sizet(matrix_host.size().cols()),
+        dlaf::to_sizet(matrix_host.size().rows()),
+        internal::hdf5_datatype<T>::dims,
+    };
+    H5::DataSpace dataspace_file(3, dims_file);
+
+    // TODO it might be needed to wait all pika tasks
```

Comment:
I've found this reference https://support.hdfgroup.org/HDF5/doc/RM/CollectiveCalls.html

AFAIU the points we should care about because they use MPI collectives internally are:
- file creation
- dataset creation

Reply:
@msimberg for the moment I didn't "guarded" above calls with the pika call to wait all tasks.

Do you think it can be problematic? Those calls happen at schedule time at the moment.

Reply:
I'm quite certain they can be problematic, for the same reasons that all other blocking calls can be problematic. I assume you haven't had problems with this in practice which is why you've left it out now? We can always see how far we get without the `wait` and add it in later if it seems ok for now.